ksmt-cvc5/src/main/kotlin/org.ksmt.solver.cvc5/KCvc5Solver.kt

```diff
@@ -0,0 +1,186 @@
+
+    override fun configure(configurator: KCvc5SolverConfiguration.() -> Unit) {
+        KCvc5SolverConfigurationImpl(solver).configurator()
+    }
+
+    override fun assert(expr: KExpr<KBoolSort>) = cvc5Try {
+        val cvc5Expr = with(exprInternalizer) { expr.internalizeExpr() }
+        currentLevelAssertions.add(expr)
+
+        solver.assertFormula(cvc5Expr)
+    }
+
+    override fun assertAndTrack(expr: KExpr<KBoolSort>): KExpr<KBoolSort> = cvc5Try {
+        val trackVar = with(ctx) { boolSort.mkFreshConst("track") }
+        val trackedExpr = with(ctx) { trackVar implies expr }
+
+        val cvc5TrackVar = with(exprInternalizer) { trackVar.internalizeExpr() }
+        val cvc5TrackedExpr = with(exprInternalizer) { trackedExpr.internalizeExpr() }
+
+        cvc5CurrentLevelTrackedAssertions.add(cvc5TrackVar)
+        currentLevelAssertions.add(trackedExpr)
+
+        solver.assertFormula(cvc5TrackedExpr)
+        solver.assertFormula(cvc5TrackVar)
+
+        trackVar
+    }
+
+    override fun push() = solver.push().also {
+        cvc5CurrentLevelTrackedAssertions = mutableListOf()
+        cvc5TrackedAssertions.add(cvc5CurrentLevelTrackedAssertions)
+
+        currentLevelAssertions = mutableListOf()
+        assertions.add(currentLevelAssertions)
+    }
+
+    override fun pop(n: UInt) {
+        require(n <= currentScope) {
+            "Can not pop $n scope levels because current scope level is $currentScope"
+        }
+
+        if (n == 0u) return
+
+        repeat(n.toInt()) {
+            cvc5TrackedAssertions.removeLast()
+            assertions.removeLast()
+        }
+        cvc5CurrentLevelTrackedAssertions = cvc5TrackedAssertions.last()
+        currentLevelAssertions = assertions.last()
+
+        solver.pop(n.toInt())
+    }
+
+    override fun check(timeout: Duration): KSolverStatus = cvc5Try {
+        cvc5LastAssumptions = TreeSet()
+        solver.updateTimeout(timeout)
+        solver.checkSat().processCheckResult()
+    }
+
+    override fun reasonOfUnknown(): String = cvc5Try {
+        require(lastCheckStatus == KSolverStatus.UNKNOWN) { "Unknown reason is only available after UNKNOWN checks" }
+        cvc5LastUnknownExplanation?.name ?: "no explanation"
+    }
+
+    override fun checkWithAssumptions(assumptions: List<KExpr<KBoolSort>>, timeout: Duration): KSolverStatus = cvc5Try {
+        val cvc5Assumptions = with(exprInternalizer) { assumptions.map { it.internalizeExpr() } }.toTypedArray()
+        cvc5LastAssumptions = cvc5Assumptions.toCollection(TreeSet())
+        solver.updateTimeout(timeout)
+        solver.checkSatAssuming(cvc5Assumptions).processCheckResult()
+    }
+
+    override fun model(): KModel = cvc5Try {
+        require(lastCheckStatus == KSolverStatus.SAT) { "Models are only available after SAT checks" }
+        return KCvc5Model(ctx, cvc5Ctx, assertions.flatten(), exprConverter, exprInternalizer)
```

Comment:
It is better to track declarations incrementally during internalization

Reply:
Can we take declarations from keys of `private val decls = HashMap<KDecl<*>, Term>()` in `KCvc5Context`?
As for uninterpreted sorts, we can collect uninterpreted sorts in `KCvc5SortInternalizer` and store them in `KCvc5Context` like this:

```
override fun visit(sort: KUninterpretedSort): Sort = cvc5Ctx.internalizeSort(sort) {
        cvc5Ctx.uninterpretedSorts += sort
        nSolver.mkUninterpretedSort(sort.name)
    }
```